src/router/index.ts

```diff
@@ -0,0 +1,422 @@
+  } else if (route.type === "projects") {
+    let hostPrefix;
+    if (route.params.profile) {
+      hostPrefix = `/${route.params.profile}`;
+    } else {
+      hostPrefix = `/seeds/${route.params.seed}`;
+    }
+
+    const content = route.params.content ? `/${route.params.content}` : "";
+
+    let peer = "";
+    if (route.params.peer) {
+      peer = `/remotes/${route.params.peer}`;
+    }
+
+    let suffix = "";
+    if (!route.params.route) {
+      if (route.params.revision) {
+        suffix = `/${route.params.revision}`;
+      }
+      if (route.params.path && route.params.path !== "/") {
+        suffix += `/${route.params.path}`;
+      }
+      if (route.params.hash) {
+        suffix += `#${route.params.hash}`;
+      }
+      if (route.params.search) {
+        suffix += `${route.params.search}`;
+      }
+    } else {
+      suffix = `/${route.params.route}`;
+      if (route.params.search) {
+        suffix += `${route.params.search}`;
+      }
+      if (route.params.hash) {
+        suffix += `#${route.params.hash}`;
+      }
+    }
+
+    if (route.params.content === "tree") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}${suffix}`;
+    } else if (route.params.content === "commits") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}${suffix}`;
+    } else if (route.params.content === "commit") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}${suffix}`;
+    } else if (route.params.content === "patches") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}${suffix}`;
+    } else if (route.params.content === "patch") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}/${route.params.patch}`;
+    } else if (route.params.content === "issue") {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}/${route.params.issue}`;
+    } else {
+      return `${hostPrefix}/${route.params.urn}${peer}${content}`;
+    }
+  } else if (
+    route.type === "registration" &&
+    route.params.activeView.type === "validateName"
+  ) {
+    return `/registration`;
+  } else if (
+    route.type === "registration" &&
+    route.params.activeView.type === "view"
+  ) {
+    return `/registration/${route.params.activeView.params.nameOrDomain}?retry=${route.params.activeView.params.retry}`;
+  } else if (
+    route.type === "registration" &&
+    (route.params.activeView.type === "checkNameAvailability" ||
+      route.params.activeView.type === "register")
+  ) {
+    if (route.params.activeView.params.owner) {
+      return `/registration/${route.params.activeView.params.nameOrDomain}/${route.params.activeView.type}?owner=${route.params.activeView.params.owner}`;
+    }
+    return `/registration/${route.params.activeView.params.nameOrDomain}/${route.params.activeView.type}`;
+  } else if (route.type === "profile") {
+    return `/${route.params.addressOrName}`;
+  } else if (route.type === "404") {
+    return null;
+  }
```

Comment:
We should be using an exhaustive check here. This `if` statement will still compile if we forgot to handle a route..

Reply:
dc69cdae858cfdff0c53f7b50ad0b16edeb67afe moves the `return null` into an else branch, so we handle the case of the forgotten route.

Reply:
b2df958c17e03ad346900a8ca6491aefd3ef3e58 ok this one is even better as exhaustive check

Reply:
00a4bf563db83d86416e145453c79b625f3987e1 and more exhaustivness 😅